### PR TITLE
Introduce encoding factory in open

### PR DESCRIPTION
**Pull request by urso**
Introduce encoding factory to determine encoding from file on open and provide
utf16 encodings with BOM support checking first 2 bytes from file.

### CODE REVIEW

**Comment by urso** [comment]
This PR makes the utf16 bom based encodings available again. See #205.

Limitations:
- files requires at least 2 bytes to be written. Otherwise open fails and harvester will not start. 
- BOM support requires seekable source => no love for stdin.

File: harvester/log.go [hunk 1]
```diff
@@ -230,12 +230,32 @@ func (h *Harvester) open() error {
...
+
+	encoding, err := h.encoding(h.file)
+	if err != nil {
+		h.file.Close()
+		return nil, err
 	}
 
-	h.setFileOffset()
+	// update file offset
+	if h.Offset > 0 {
```
Comment (ruflin) [inline harvester/log.go]: I would still keep this inside the setFileOffset method

**Comment by ruflin** [comment]
@urso It seems to break the appveyor tests ...

**Comment by ruflin** [comment]
@urso about the stdin issue. Stdin should NOT be part of the log harvester. I'm aware that it still inside but we should extract it later as an independent type so the above will not be an issue.

**Comment by urso** [comment]
I see this failing badly with stdin in general. Needs some more thinking

File: harvester/encoding/utf16.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	// read Byte Order Marker (BOM)
+	var buf [2]byte
+	n, err := in.Read(buf[:])
+	if err != nil {
+		in.Seek(offset, os.SEEK_SET)
+		return nil, err
+	}
+	if n < 2 {
+		in.Seek(offset, os.SEEK_SET)
+		return nil, transform.ErrShortSrc
```
Comment (ruflin) [inline harvester/encoding/utf16.go]: Lets add a Info message here.

File: harvester/encoding/utf16.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	switch {
+	case buf[0] == 0xfe && buf[1] == 0xff:
+		inEndiannes = bigEndian
+	case buf[0] == 0xff && buf[1] == 0xfe:
+		inEndiannes = littleEndian
+	}
+
+	// restore offset if BOM is missing or this function was not
+	// called with read pointer at beginning of file
+	if !keepOffset || inEndiannes == unknownEndianess {
```
Comment (ruflin) [inline harvester/encoding/utf16.go]: This basically mean, it is probably not utf16, right?

File: harvester/encoding/utf16.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+		in, ok := in_.(io.ReadSeeker)
+		if !ok {
+			return nil, ErrUnsupportedSourceTypeBOM
+		}
+
+		return utf16Seekable(in, e)
+	}
+}
+
+func utf16Seekable(in io.ReadSeeker, endianness endianness) (Encoding, error) {
```
Comment (ruflin) [inline harvester/encoding/utf16.go]: Would it solve your problem, if filebeat wouldn't ever start reading empty files?

File: harvester/log.go [hunk 5]
```diff
@@ -152,67 +155,88 @@ func (h *Harvester) backOff() {
...
-// Sets the offset of the file to the right place. Takes configuration options into account
-func (h *Harvester) setFileOffset() {
-	if h.Offset > 0 {
-		h.file.Seek(h.Offset, os.SEEK_SET)
-	} else if h.Config.TailFiles {
-		h.file.Seek(0, os.SEEK_END)
-	} else {
-		h.file.Seek(0, os.SEEK_SET)
-	}
+func (h *Harvester) openStdin() (encoding.Encoding, error) {
```
Comment (ruflin) [inline harvester/log.go]: :+1:

**Comment by ruflin** [comment]
In general looks good to me. To be honest it is very hard to figure out if the all the checks are still made in the identical order as they were made before as quite a few parts were moved around. But as our tests cover most of the edge cases, I'm quite confident that this will work as expected. Good for me to move forward.

File: harvester/encoding/utf16.go (hunk 4 above)
Comment (urso) [inline harvester/encoding/utf16.go]: this would help. On the other hand check opening loop. In case of file being empty, the loop will retry after backing off. For now it's an okish solution I would say.

File: harvester/encoding/utf16.go (hunk 3 above)
Comment (urso) [inline harvester/encoding/utf16.go]: No, problem is the BOM is optional. Missing BOM does not say anything about the encoding. If missing the codec must fallback to either big endian or little endian. Default fallback should be big endian, but windows uses little endian by default.

File: harvester/encoding/utf16.go (hunk 2 above)
Comment (urso) [inline harvester/encoding/utf16.go]: +1

Comment (urso) [inline harvester/encoding/utf16.go]: added info to open function

File: harvester/encoding/utf16.go (hunk 3 above)
Comment (ruflin) [inline harvester/encoding/utf16.go]: As it seems like we can't guarantee to detect it correctly, should we make it configurable?

Comment (urso) [inline harvester/encoding/utf16.go]: make what configurable? It already is. There are 3 codecs: utf-16be-bom (fallback to BigEndian), utf-16le-bom (fallback to LittleEndian) and utf-16-bom (fail if BOM is missing).
